include/lbann/data_readers/sample_list_jag.hpp

```diff
@@ -112,53 +98,144 @@ class sample_list_jag {
   /// Check if a sample index is in the valid range
   bool check_index(size_t idx) const;
 
-  /// Serialize sample list for a partition
-  bool to_string(size_t p, std::string& sstr) const;
-
-  /// Serialize sample list for all partitions
+  /// Serialize sample list
   bool to_string(std::string& sstr) const;
 
-  /// Write the sample list of partition p
-  void write(size_t p, const std::string filename) const;
-
-  /// Write the sample list of each partitions
+  /// Write the sample list
   void write(const std::string filename) const;
 
   /// Allow read-only access to the internal list data
   const samples_t& get_list() const;
 
-  /// Copy the internal list data for partition p
-  bool get_list(size_t p, samples_t& l_p) const;
-
-  /// Allow read-only access to the internal list data for partition p via iterators
-  std::pair<samples_t::const_iterator, samples_t::const_iterator> get_list(size_t p) const;
-
   /// Allow the read-only access to the list header
   const sample_list_header& get_header() const;
 
   /// Allow read-only access to the metadata of the idx-th sample in the list
   const sample_t& operator[](size_t idx) const;
 
-  const std::string& get_samples_filename(sample_id_t id) const {
-    return (m_sample_id_map[id]).first;
+  const std::string& get_samples_filename(sample_file_id_t id) const {
+    return std::get<0>(m_file_id_stats_map[id]);
+  }
+
+  const std::string& get_samples_dirname() const {
+    return m_header.get_file_dir();
   }
 
-  hid_t get_samples_hdf5_handle(sample_id_t id) const {
-    return (m_sample_id_map[id]).second;
+  hid_t get_samples_hdf5_handle(sample_file_id_t id) const {
+    hid_t h = std::get<1>(m_file_id_stats_map[id]);
+    return h;
   }
 
-  void set_samples_filename(sample_id_t id, const std::string& filename) {
-    m_sample_id_map[id].first = filename;
+  void set_samples_filename(sample_file_id_t id, const std::string& filename) {
+    std::get<0>(m_file_id_stats_map[id]) = filename;
   }
 
-  void set_samples_hdf5_handle(sample_id_t id, hid_t h) {
-    m_sample_id_map[id].second = h;
+  void set_files_hdf5_handle(const std::string& filename, hid_t h) {
+    sample_file_id_t id = 0;
+    for (auto&& e : m_file_id_stats_map) {
```

Comment:
As a performance optimization, we can use another map from the filename to sample_file_id_t.
If maintaining two different data structures consistently is a burden, there is a container called boos::multi_index.